eZ/Bundle/EzPublishCoreBundle/EventListener/ContentDownloadRouteReferenceListener.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Bundle\EzPublishCoreBundle\EventListener;
+
+use eZ\Publish\API\Repository\Values\Content\Field;
+use eZ\Publish\Core\MVC\Symfony\Event\RouteReferenceGenerationEvent;
+use eZ\Publish\Core\MVC\Symfony\MVCEvents;
+use InvalidArgumentException;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ContentDownloadRouteReferenceListener implements EventSubscriberInterface
+{
+    const ROUTE_NAME = 'ez_content_download';
+    const OPT_FIELD_IDENTIFIER = 'fieldIdentifier';
+    const OPT_CONTENT = 'content';
+    const OPT_CONTENT_ID = 'contentId';
+    const OPT_DOWNLOAD_NAME = 'filename';
+    const OPT_LANGUAGE = 'language';
+    const OPT_VERSION = 'version';
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            MVCEvents::ROUTE_REFERENCE_GENERATION => 'onRouteReferenceGeneration'
+        );
+    }
+
+    /**
+     * @throws \InvalidArgumentException If the required arguments are not correct
+     */
+    public function onRouteReferenceGeneration( RouteReferenceGenerationEvent $event )
+    {
+        $routeReference = $event->getRouteReference();
+
+        if ( $routeReference->getRoute() != self::ROUTE_NAME )
+        {
+            return;
+        }
+
+        $options = $routeReference->getParams();
+
+        $resolver = new OptionsResolver();
+        $this->configureOptions( $resolver );
+        $options = $resolver->resolve( $options );
+
+        if ( isset( $options[self::OPT_LANGUAGE]) )
+        {
+            $routeReference->set( self::OPT_LANGUAGE, $options[self::OPT_LANGUAGE] );
+        }
+
+        if ( isset( $options[self::OPT_VERSION] ) )
+        {
+            $routeReference->set( self::OPT_VERSION, $options[self::OPT_VERSION] );
+        }
+
+        $routeReference->set( self::OPT_CONTENT_ID, $options[self::OPT_CONTENT_ID] );
+        $routeReference->set( self::OPT_DOWNLOAD_NAME, $options[self::OPT_DOWNLOAD_NAME] );
+    }
+
+    /**
+     * @param $resolver
+     */
+    protected function configureOptions( OptionsResolver $resolver )
+    {
+        $resolver->setRequired( [ self::OPT_CONTENT, self::OPT_FIELD_IDENTIFIER ] );
+        $resolver->setDefined(
+            [ self::OPT_VERSION, self::OPT_LANGUAGE, self::OPT_DOWNLOAD_NAME, self::OPT_CONTENT_ID ]
```

Comment:
`self::OPT_DOWNLOAD_NAME` is not needed here since you set a default value for it below. Same for `self::OPT_CONTENT_ID`